src/services/preProcess.ts

```diff
@@ -16,19 +16,31 @@ namespace ts {
         let lastToken: SyntaxKind;
         let currentToken: SyntaxKind;
         let braceNesting = 0;
+        let templateNesting = 0;
         // assume that text represent an external module if it contains at least one top level import/export
         // ambient modules that are found inside external modules are interpreted as module augmentations
         let externalModule = false;
 
         function nextToken() {
             lastToken = currentToken;
             currentToken = scanner.scan();
-            if (currentToken === SyntaxKind.OpenBraceToken) {
+            if (currentToken === SyntaxKind.CloseBraceToken && templateNesting > 0) {
+                currentToken = scanner.reScanTemplateToken();
+            }
+
+            if (currentToken === SyntaxKind.TemplateHead) {

```

Comment:
This _technically_ needs to track more than one scanning context. For example,
```
`${(<div>This is JSX text so these: `` ${} mean nothing, and in fact, the {} are
(empty) interpolation areas! So ${import("mod")} is a dynamic import! Also,
you could have ${/*A comment*/} or ${/*import("ignored")*/} a commented module </div>)}`;
```

The textmate grammar chokes on that one, too, since it involves multiple nested scanning contexts, but it is parsed correctly.

Reply:
I am a bit unfamiliar with how this works, but I will add these test cases and see if I can figure it out. Thanks!